lib/Models/UserDrawing.ts

```diff
@@ -0,0 +1,434 @@
+
+    // helper for dragging points around
+    this.dragHelper = new DragPoints(options.terria, customDataSource => {
+      if (this.onPointMoved) {
+        this.onPointMoved(customDataSource);
+      }
+      this.prepareToAddNewPoint();
+    });
+  }
+
+  @computed get mapItems() {
+    return [this.pointEntities, this.otherEntities];
+  }
+
+  @computed get svgPoint() {
+    /**
+     * SVG element for point drawn when user clicks.
+     * http://stackoverflow.com/questions/24869733/how-to-draw-custom-dynamic-billboards-in-cesium-js
+     */
+    var svgDataDeclare = "data:image/svg+xml,";
+    var svgPrefix =
+      '<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" width="20px" height="20px" xml:space="preserve">';
+    var svgCircle =
+      '<circle cx="10" cy="10" r="5" stroke="rgb(0,170,215)" stroke-width="4" fill="white" /> ';
+    var svgSuffix = "</svg>";
+    var svgString = svgPrefix + svgCircle + svgSuffix;
+
+    // create the cesium entity
+    return svgDataDeclare + svgString;
+  }
+
+  enterDrawMode() {
+    this.dragHelper.setUp();
+
+    // If we have finished a polygon, don't allow more points to be drawn. In future, perhaps support multiple polygons.
+    if (this.inDrawMode || this.closeLoop) {
+      // Do nothing
+      return;
+    }
+
+    this.inDrawMode = true;
+
+    if (isDefined(this.terria.cesium)) {
+      this.terria.cesium.cesiumWidget.canvas.setAttribute(
+        "style",
+        "cursor: crosshair"
+      );
+    } else if (isDefined(this.terria.leaflet)) {
+      const container = document.getElementById("cesiumContainer");
+      if (container !== null) {
+        container.setAttribute("style", "cursor: crosshair");
+      }
+    }
+
+    // Cancel any feature picking already in progress.
+    runInAction(() => {
+      this.terria.pickedFeatures = undefined;
+    });
+    const that = this;
+
+    // Line will show up once user has drawn some points. Vertices of line are user points.
+    this.otherEntities.entities.add(<any>{
+      name: "Line",
+      polyline: <any>{
+        positions: new CallbackProperty(function() {
+          const pos = that.getPointsForShape();
+          if (isDefined(pos) && that.closeLoop) {
+            pos.push(pos[0]);
+          }
+          return pos;
+        }, false),
+
+        material: new PolylineGlowMaterialProperty(<any>{
+          color: new Color(0.0, 0.0, 0.0, 0.1),
+          glowPower: 0.25
+        }),
+        width: 20
+      }
+    });
+
+    this.terria.overlays.add(this);
+
+    // Listen for user clicks on map
+    const pickPointMode = new MapInteractionMode({
+      message: this.getDialogMessage(),
+      buttonText: this.getButtonText(),
+      onCancel: () => {
+        this.terria.mapInteractionModeStack.pop();
+        this.cleanUp();
+      }
+    });
+
+    this.terria.mapInteractionModeStack.push(pickPointMode);
+    autorun(async reaction => {
```

Comment:
Is it possible to cancel picking without having this reaction disposed?

Reply:
Thanks!

I have modified the code to slightly improve clarity & dispose the reaction when the interaction is cancelled.